typing/mode.ml

```diff
@@ -1592,6 +1589,249 @@ type 'a comonadic_with = 'a C.comonadic_with =
+    | Max_with m_ax, ax | Min_with m_ax, ax -> (
+      match Axis.eq m_ax ax with
+      | None -> NoneResponsible
+      | Some Refl -> SourceIsSingle)
+    | Monadic_to_comonadic_min, ax -> handle_monadic_to_comonadic ax
+    | Monadic_to_comonadic_max, ax ->
+      (* Can't use an "or" pattern here due to GADT matching limitations,
+         so I've just made a separate function for code neatness, then inlined it *)
+      handle_monadic_to_comonadic ax
+    | Comonadic_to_monadic _, ax -> (
+      (* See [Lattices_mono.monadic_to_comonadic_min] for why these are as they are *)
+      match ax with
+      | Uniqueness -> Axis Linearity
+      | Contention -> Axis Portability
+      | Visibility -> Axis Statefulness)
+    | Proj _, _
+    | Local_to_regional, _
+    | Regional_to_local, _
+    | Locality_as_regionality, _
+    | Regional_to_global, _
+    | Global_to_regional, _ ->
+      .
+
+(** Container for an adjoint function. We use a special record type instead of inlining
+the type so that we can have universal quantification over the input object types. *)
+type ('l1, 'r1, 'l2, 'r2) shint_to_axhint_side =
+  | LeftAdjoint : (_, allowed, allowed, disallowed) shint_to_axhint_side
+  | RightAdjoint : (allowed, _, disallowed, allowed) shint_to_axhint_side
+
+let shint_to_axhint_side_fn :
+    type a b l1 l2 r1 r2.
+    (l1, r1, l2, r2) shint_to_axhint_side ->
+    b C.obj ->
+    (a, b, l1 * r1) C.morph ->
+    (b, a, l2 * r2) C.morph =
+  fun (type a b l1 l2 r1 r2) (side : (l1, r1, l2, r2) shint_to_axhint_side) :
+      (b C.obj -> (a, b, l1 * r1) C.morph -> (b, a, l2 * r2) C.morph) ->
+   match side with
+   | LeftAdjoint -> C.left_adjoint
+   | RightAdjoint -> C.right_adjoint
+
+let shint_to_axhint_side_le :
+    type a l1 l2 r1 r2.
+    (l1, r1, l2, r2) shint_to_axhint_side -> a C.obj -> a -> a -> bool =
+  fun (type a l1 l2 r1 r2) (side : (l1, r1, l2, r2) shint_to_axhint_side)
+      (a_obj : a C.obj) x y ->
+   match side with
+   | LeftAdjoint -> C.le a_obj x y
+   | RightAdjoint -> C.le a_obj y x
+
+let rec shint_to_axhint :
+    type r a left1 left2 right1 right2.
+    r Lattices_mono.obj ->
+    r ->
+    (r, left1 * right1) S.hint ->
+    (r, a) Axis.t ->
+    (left1, right1, left2, right2) shint_to_axhint_side ->
+    (a, Hint.morph, Hint.const) axhint =
+  let open Lattices_mono in
+  fun (type r a) (r_obj : r obj) (r : r) (r_shint : (r, left1 * right1) S.hint)
+      (ax : (r, a) Axis.t) side ->
+    (* This function is for when we have a solver hint (aka "shint") for a product lattice and
+       wish to project the hint to be for a single axis and convert it to a mode "axhint". *)
+    let a_obj = proj_obj ax r_obj in
+    let proj = Proj (r_obj, ax) in
+    let a = apply a_obj proj r in
+    let a_shint : (a, left1 * right1) S.hint =
+      let rec compose_shint_with_proj = function
+        | S.Morph (morph_hint, morph, b_shint) ->
+          S.Morph (morph_hint, compose a_obj proj morph, b_shint)
```

Comment:
Instead of calling `compose`, simply pass additional argument `ax` to `single_axis_shint_to_axhint`.
After all, the meaning of `Proj` is already captured by `find_responsible_axis_prod`?